### PR TITLE
Feature - Added in tag parent id filter

### PR DESCRIPTION
**Pull request by tonypartridge**
Pull Request for Issue #19101 .

### Summary of Changes

Added Tag Parent filter in tags view.

### Testing Instructions

Have nested tags, filter by parent all tags below should show. Also test max levels, parent tag should by level 1 and so on. 

@richard67 @ggppdk @Septdir @rgblogs can we get some tests please :-)

### CODE REVIEW

**Comment by Septdir** [comment]
I have tested this item :white_check_mark: successfully on 7cabf95c7db8f8882769121138dbd44ab5e0e93a<br /><br />Works good.But I do not like how select looks
It is possible not to get out the name of the parent but just use it `-`
For example
```
Level 1
- Level 2
-- Level 3
- Level 2
Level 1
```
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22322">issues.joomla.org/tracker/joomla-cms/22322</a>.</sub>

File: administrator/language/en-GB/en-GB.com_tags.ini [hunk 1]
```diff
@@ -90,6 +90,8 @@ COM_TAGS_FIELDSET_TAGGED_ITEMS="Tagged Items"
 COM_TAGS_FIELDSET_URLS_AND_IMAGES="Links and Images"
 COM_TAGS_FILTER_SEARCH_DESC="Search in tag title, alias and note. Prefix with ID: to search for a tag ID."
 COM_TAGS_FILTER_SEARCH_LABEL="Search Tags"
+COM_TAGS_FILTER_TAGPARENT_DESC="Filters the tag items list where the parent is this selected tag"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_tags.ini]: Add a full stop please

File: administrator/language/en-GB/en-GB.com_tags.ini [hunk 2]
```diff
@@ -136,6 +138,7 @@ COM_TAGS_SAVE_SUCCESS="Tag saved."
 COM_TAGS_SEARCH_TYPE_DESC="All will return items that have all of the tags. Any will return items that have at least one of the tags."
 COM_TAGS_SEARCH_TYPE_LABEL="Match Type"
 COM_TAGS_SELECT_TAGTYPE="- Select Tag Type -"
+COM_TAGS_SELECT_TAGPARENT="- Select Tag Parent -"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_tags.ini:141]: I don't mind if it is tag parent or parent tag but please be consistent

**Review by brianteeman — CHANGES_REQUESTED** [review]

**Comment by ggppdk** [comment]
This needs a little more work
- the max levels filter is applied twice once according to parent menu item, and once according to top-most level

please remove the existing code of 
```php
		// Filter on the level.
		if ($level = $this->getState('filter.level'))
		{
			$query->where('a.level <= ' . (int) $level);
		}
```
and append it as an ELSEIF to the parent ID filter:
```php
		if (!empty($parentId))
		{
				// ... this is correct
				// here the max level filter will be applied according to currently chosen parent id
		}

		// Filter on the level.
		elseif ($level = $this->getState('filter.level'))
		{
			$query->where('a.level <= ' . (int) $level);
		}
```

**Comment by tonypartridge** [comment]
@ggppdk  of course it does, thanks I've updated it now.

File: administrator/language/en-GB/en-GB.com_tags.ini (hunk 2 above)
Comment (Quy) [reply administrator/language/en-GB/en-GB.com_tags.ini:141]: Sort key in alpha

**Comment by tonypartridge** [comment]
@Septdir @ggppdk would you be kind enough to test and let’s get this ticked off?

**Comment by Septdir** [comment]
I have tested this item :white_check_mark: successfully on f37b10b90053effca747aac920c95a489d35ee92<br /><br />Test on Joomla! 3.9.0 RC 2

P.S Maybe the parent tag itself should also be displayed after filtering?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22322">issues.joomla.org/tracker/joomla-cms/22322</a>.</sub>

**Comment by rogercreagh** [comment]
The filter field is using the type=tag. This will give a list of all tags, which is not helpful if you have a few hundred tags.
Selecting a tag which has no children as the parent filter is meaningless.
A better solution might be to only show tags that are parents in the filter dropdown.
eg:
```
<field
	name="parent_id"
	label="Parent"
	description="Select parent tag to filter by"
	onchange="this.form.submit();"
	default=""
	query="SELECT DISTINCT c.`parent_id`, p.id AS value, p.title AS parent, p.published 
        	FROM `#__tags` AS c 
		LEFT JOIN `#__tags` AS p ON c.`parent_id` = p.`id` 
		WHERE p.published
		ORDER BY p.lft"
	>
	<option value="">Select parent</option>
</field>
```
Obviously you need language strings for the Label, Description and Default value, but I think you have those.

Also I find it is very helpful to have the parent tag visible at the top of the list so I use
```
// Filter on the level.
if (($level = $this->getState('filter.level')) AND !($parent = $this->getState('filter.parent')))
// if we have set a parent then don't do the simple level filter
{
    $query->where('a.level <= ' . (int) $level);
}
// Filter on the parent.
if ($parent = $this->getState('filter.parent'))
{
    if (($level>1) AND ($lim = self::getPlftPrgt($parent))) { //if level is empty or 1 then only show the direct children 
        $query->where('a.lft BETWEEN '.$lim);
    } else {
        $query->where('a.parent_id = ' . (int) $parent. ' OR a.id = '.(int) $parent );
        //the OR ensures that the selected parent is show if level is not set
    }		   		    
}
```
where ```getPlftPrgt($parent)``` is a function to return the lft and rgt limits as a string eg "97 AND 101" rather than use a subquery
```
protected function getPlftPrgt($pid)
{
    $db1 = JFactory::getDBO();
    $qry = 'SELECT p.lft, p.rgt FROM #__tags AS p WHERE p.id = '.$pid;
    $db1->setQuery($qry);
    $ans = $db1->loadObject();	    
    if (!$db1->query()) {
    //    throw new Exception($db->getErrorMsg(), 500); //could provide a better error trap than simply ignoring it
        return false;
    }
    if (isset($ans->lft) && isset($ans->rgt)) {
       return $ans->lft.' AND '.$ans->rgt;
    }
    return false;
}
```
I implemented this on my sites a while back but never realised it was a feature request - sorry about that.
This version addresses the issue of the selector for the filter getting cumbersome if you have many tags and also ensures that the parent is always show at the top of the list<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22322">issues.joomla.org/tracker/joomla-cms/22322</a>.</sub>

**Comment by kofaysi** [comment]
I have tested this item :white_check_mark: successfully on f37b10b90053effca747aac920c95a489d35ee92<br /><br />J3.9.2

I think it works as stated and expected.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22322">issues.joomla.org/tracker/joomla-cms/22322</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on f37b10b90053effca747aac920c95a489d35ee92<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22322">issues.joomla.org/tracker/joomla-cms/22322</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22322">issues.joomla.org/tracker/joomla-cms/22322</a>.</sub>

**Comment by HLeithner** [comment]
Because this PR is a new feature it can't be merged into 3.9, please create a new PR for Joomla! 4.

**Comment by tonypartridge** [comment]
There will be 3.10 will there not?

**Comment by HLeithner** [comment]
3.10 will not have new features. It's a compatibility version for upgrading to Joomla 4.0

**Comment by viocassel** [comment]
😞

**Pull request closed without merging**